File: 11.java
// LeetCode 11 Container with Most Water
// Reference: https://leetcode.com/problems/container-with-most-water/solutions/3701708/best-method-c-java-python-beginner-friendly/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

class Solution {
    public int maxArea(int[] height) {
        // Two pointers
        int left = 0;
        int right = height.length - 1;
        // max area = most water
        int maxArea = 0;

        while (left < right){
            int currentArea = Math.min(height[left], height[right]) * (right - left);
            maxArea = Math.max(maxArea, currentArea);

            if (height[left] < height[right]){
                left --;
            } else {
                right --;
            }
        }
        return maxArea;
    }
}

File: 11.js
// LeetCode 11 Container with Most Water
// Reference: https://leetcode.com/problems/container-with-most-water/?envType=study-plan-v2&envId=top-interview-150

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    // Two pointers
    let result = 0, left = 0, right = height.length - 1;

    while (left < right){
        let smallestSide = Math.min(height[left], height[right]);
        let area = (right - left) * smallestSide;

        if (area > result) result = area;

        if (height[left] < height[right]) left++;
        else right--;
    }
    
    return result;
};

